File: 백준/Silver/2468. 안전 영역/안전 영역.cc
#include<bits/stdc++.h>
using namespace std;
int n, cnt, ret = 1;
int a[101][101], visited[101][101];
int dy[4] = {-1, 0, 1, 0};
int dx[4] = {0, 1, 0, -1};
void dfs(int y, int x, int k){
    visited[y][x] = 1;
    for(int i = 0; i < 4; i++){
        int ny = y + dy[i];
        int nx = x + dx[i];
        if(ny < 0 || nx < 0 || ny >= n || nx >= n) continue;
        if(a[ny][nx] > k && !visited[ny][nx]) dfs(ny, nx, k);
    }
    return;
}
int main(){
    cin >> n;
    for(int i = 0; i < n; i++){
        for(int j = 0; j < n; j++){
            cin >> a[i][j];
        }
    }
    for(int k = 1; k < 101; k++){
        fill(&visited[0][0], &visited[0][0] + 101 * 101, 0);
        int cnt = 0;
        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                if(a[i][j] > k && !visited[i][j]){
                    dfs(i, j, k); cnt++;
                }
            }
        }
        ret = max(ret, cnt);
    }
    cout << ret << '\n';
}

File: 백준/Silver/2468. 안전 영역/안전 영역.java
import java.io.*;
import java.util.*;

public class Main {
    static class Pair{
        int y; int x;
        Pair(int y, int x){
            this.y = y;
            this.x = x;
        }
    }
    static int[][] board;
    static boolean[][] visited;
    static int n;
    static int[] dy = {-1, 0, 1, 0}, dx = {0, 1, 0, -1};
    static void dfs(int y, int x, int h){
        var stk = new ArrayDeque<Pair>();
        stk.push(new Pair(y, x));
        while(!stk.isEmpty()){
            Pair p = stk.pop();
            int cy = p.y;
            int cx = p.x;
            for(int d = 0; d < 4; d++){
                int ny = cy + dy[d];
                int nx = cx + dx[d];
                if(ny < 0 || nx < 0 || ny >= n || nx >= n) continue;
                if(visited[ny][nx] || board[ny][nx] <= h) continue;
                visited[ny][nx] = true;
                stk.push(new Pair(ny, nx));
            }
        }
    }
    public static void main(String[] args) throws IOException {
        var in = new BufferedReader(new InputStreamReader(System.in));
        var out = new PrintWriter(System.out);

        n = Integer.parseInt(in.readLine());
        board = new int[n][n];
        for(int i = 0; i < n; i++){
            var st = new StringTokenizer(in.readLine());
            for(int j = 0; j < n; j++){
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        int ret = -1;
        for(int h = 0; h <= 100; h++){
            visited = new boolean[n][n];
            int cnt = 0;
            for(int i = 0; i < n; i++){
                for(int j = 0; j < n; j++){
                    if(board[i][j] <= h || visited[i][j]) continue;
                    dfs(i, j, h); cnt++;
                }
            }
            ret = Math.max(ret, cnt);
        }
        out.println(ret);

        out.flush();
        out.close();
    }
}